packages/flutter_tools/lib/src/build_system/targets/custom_device.dart

```diff
@@ -0,0 +1,169 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import '../../artifacts.dart';
+import '../../base/file_system.dart';
+import '../../build_info.dart';
+import '../../devfs.dart';
+import '../../project.dart';
+import '../build_system.dart';
+import '../depfile.dart';
+import '../exceptions.dart';
+import 'assets.dart';
+import 'common.dart';
+import 'icon_tree_shaker.dart';
+
+/// Creates a bundle for the custom device target.
+abstract class BundleCustomDeviceAssets extends Target {
+  const BundleCustomDeviceAssets(this.targetPlatform);
+
+  factory BundleCustomDeviceAssets.forBuildMode(BuildMode mode, TargetPlatform targetPlatform) {
+    switch (mode) {
+      case BuildMode.debug:
+        return DebugCustomDeviceAssets(targetPlatform);
+      case BuildMode.profile:
+       return ProfileCustomDeviceAssets(targetPlatform);
+      default:
+      throw FallThroughError();
+    }
+  }
+
+  final TargetPlatform targetPlatform;
+
+  @override
+  List<Target> get dependencies => <Target>[
+    const KernelSnapshot(),
+  ];
+
+  @override
+  List<Source> get inputs => const <Source>[
+    Source.pattern('{FLUTTER_ROOT}/packages/flutter_tools/lib/src/build_system/targets/custom_device.dart'),
+    Source.pattern('{PROJECT_DIR}/pubspec.yaml'),
+    ...IconTreeShaker.inputs,
+  ];
+
+  @override
+  List<String> get depfiles => const <String>[
+    'flutter_assets.d',
+  ];
+
+  @override
+  Future<void> build(Environment environment) async {
+    if (environment.defines[kBuildMode] == null) {
+      throw MissingDefineException(kBuildMode, 'bundle_custom_device_assets');
+    }
+    final BuildMode buildMode = getBuildModeForName(environment.defines[kBuildMode]!);
+    final Directory outputDirectory = environment.outputDir;
+    if (!outputDirectory.existsSync()) {
+      outputDirectory.createSync();
+    }
+
+    // Only copy the kernel blob in debug mode.
+    if (buildMode == BuildMode.debug) {
+      environment.buildDir.childFile('app.dill')
+        .copySync(outputDirectory.childFile('kernel_blob.bin').path);
+    }
+    environment.fileSystem.file(environment.artifacts.getArtifactPath(Artifact.flutterEngineLibrary)).copySync(outputDirectory.childFile('libflutter_engine.so').path);
```

Comment:
Comment regarding copying went missing @zanderso : "Is it possible to avoid copying this string here and using a constant or a lookup using Artifact.libFlutterEngine instead?" 
```suggestion
    environment.fileSystem.file(environment.artifacts.getArtifactPath(Artifact.flutterEngineLibrary)).copySync(outputDirectory.childFile(basename(environment.artifacts.getArtifactPath(Artifact.flutterEngineLibrary))).path);
```